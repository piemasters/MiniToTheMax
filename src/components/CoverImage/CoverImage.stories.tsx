import type { Meta, StoryObj } from '@storybook/react-webpack5';

import { CoverImage } from './CoverImage';

const meta: Meta<typeof CoverImage> = {
  component: CoverImage,
  argTypes: {
    image: {
      name: 'image',
      description: 'Cover image',
      control: { type: 'object' },
    },
    title: {
      name: 'title',
      description: 'The title of the post',
      defaultValue: 'This is the post title',
      control: { type: 'text' },
    },
    tall: {
      name: 'tall',
      description: 'Increase the offset of the title for taller images',
      defaultValue: false,
      control: { type: 'boolean' },
    },
  },
};

export default meta;
type Story = StoryObj<typeof CoverImage>;

export const Primary: Story = {
  args: {
    image: {
      layout: 'constrained', // "fullWidth" | "fixed";
      backgroundColor: '#585858',
      width: 1800,
      height: 900,
      images: {
        fallback: {
          src: '/images/example-cover.jpg',
          srcSet:
            '/images/example-cover.jpg 450w,\n/images/example-cover.jpg 900w,\n/images/example-cover.jpg 1800w',
          sizes: '(min-width: 1800px) 1800px, 100vw',
        },
        sources: [
          {
            srcSet:
              '/images/example-cover.jpg 450w,\n/images/example-cover.jpg 900w,\n/images/example-cover.jpg 1800w',
            type: 'image/jpg',
            sizes: '(min-width: 1800px) 1800px, 100vw',
          },
        ],
      },
    },
    title: `This is the post title`,
    tall: false,
  },
};
