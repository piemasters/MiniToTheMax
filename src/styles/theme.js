export default {
  colors: {
    text: '#000',
    muted: '#999',
  },
};
