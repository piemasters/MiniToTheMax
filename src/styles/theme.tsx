export interface Theme {
  colors: {
    text: string;
    muted: string;
  };
}

const appTheme: Theme = {
  colors: {
    text: '#000',
    muted: '#999',
  },
};

export default appTheme;
