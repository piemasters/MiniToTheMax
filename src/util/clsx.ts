/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable @typescript-eslint/no-explicit-any */
export type ClassValue =
  | ClassArray
  | ClassDictionary
  | string
  | number
  | null
  | boolean
  | undefined;
type ClassDictionary = Record<string, any>;
type ClassArray = ClassValue[];

function toVal(mix: string | number | boolean | unknown[] | ClassDictionary) {
  let k,
    y,
    str = '';

  if (typeof mix === 'string' || typeof mix === 'number') {
    str += mix;
  } else if (typeof mix === 'object') {
    if (Array.isArray(mix)) {
      const len = mix.length;
      for (k = 0; k < len; k++) {
        if (mix[k]) {
          if ((y = toVal(mix[k]))) {
            str && (str += ' ');
            str += y;
          }
        }
      }
    } else {
      for (y in mix) {
        if (mix[y]) {
          str && (str += ' ');
          str += y;
        }
      }
    }
  }

  return str;
}

export function clsx(...args: ClassValue[]): string {
  let i = 0,
    tmp,
    x,
    str = '';
  const len = args.length;
  for (; i < len; i++) {
    if ((tmp = args[i])) {
      if ((x = toVal(tmp))) {
        str && (str += ' ');
        str += x;
      }
    }
  }
  return str;
}

export default clsx;
