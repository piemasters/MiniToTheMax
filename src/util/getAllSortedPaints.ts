import { PaintDetails } from '../types';
import { textToId } from './textToId';
import { sortPaints } from './sortPaints';
import {
  airPaints,
  basePaints,
  contrastPaints,
  dryPaints,
  layerPaints,
  shadePaints,
  sprayPaints,
  technicalPaints,
} from '../data/paints/citadel';
import { gameColorPaints } from '../data/paints';

export const getAllSortedPaints = (): PaintDetails[] => {
  const allPaints: PaintDetails[] = [
    // citadel
    ...airPaints,
    ...basePaints,
    ...contrastPaints,
    ...dryPaints,
    ...layerPaints,
    ...shadePaints,
    ...sprayPaints,
    ...technicalPaints,
    // vallejo
    ...gameColorPaints,
  ];

  // Remove duplicates using a Set
  const uniquePaints = Array.from(
    new Set(
      allPaints.map(
        (paint) =>
          `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}`
      )
    )
  ).map((uniqueKey) =>
    allPaints.find(
      (paint) =>
        `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}` ===
        uniqueKey
    )
  ) as PaintDetails[];

  // Sort the paints using the sortPaints function
  return uniquePaints.sort(sortPaints);
};
