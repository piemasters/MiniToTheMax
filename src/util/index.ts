export * from './pagination';
export * from './paints';
export * from './removeCategoryName';
export * from './transitionProvider';
