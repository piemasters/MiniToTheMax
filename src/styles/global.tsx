import { css, SerializedStyles } from '@emotion/react';

import { Theme } from './theme';
import { resetStyles } from './reset';

export const globalStyles = (
  theme: Theme
): [SerializedStyles, SerializedStyles, SerializedStyles] => {
  return [resetStyles, gatsby(), blockquote(theme)];
};

const blockquote = (theme: Theme) => css`
  blockquote {
    background: ${theme.colors.lightgrey};
    border-left: 0.7rem solid ${theme.colors.primary};
    border-radius: 5px;
    margin: 1rem;
    padding: 0.8rem;
    quotes: '\\201C''\\201D''\\2018''\\2019';
    &:before {
      color: ${theme.colors.textSecondary};
      content: open-quote;
      font-size: 3rem;
      line-height: 0.1rem;
      margin-right: 1rem;
      vertical-align: -0.4em;
    }
    p {
      color: ${theme.colors.textSecondary};
      display: block;
      font-size: 0.8rem;
      font-style: italic;
      margin-top: 1rem;
      &:first-of-type {
        display: inline;
      }
    }
  }
`;

const gatsby = () => css`
  .gatsby-resp-image-wrapper {
    width: 100%;
  }
`;
