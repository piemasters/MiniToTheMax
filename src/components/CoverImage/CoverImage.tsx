import type { FC } from 'react';
import { css, useTheme } from '@emotion/react';
import { GatsbyImage, IGatsbyImageData } from 'gatsby-plugin-image';

import { Theme } from '../../styles/theme';

export interface CoverImageProps {
  image: IGatsbyImageData;
  title: string;
  tall?: boolean;
}

export const CoverImage: FC<CoverImageProps> = ({
  image,
  title,
  tall = false,
}) => {
  const theme = useTheme() as Theme;

  const headerContainer = css`
    position: relative;
  `;

  const postHeaderStyle = css`
    background-color: ${theme.colors.backgroundTransparent};
    bottom: ${tall ? '20px' : '0px'};
    color: ${theme.colors.textLight};
    font-size: 1.2rem;
    padding: 0.5rem 1rem;
    position: absolute;
    right: 1rem;
  `;

  return (
    <div css={headerContainer} data-testid="cover-image">
      <GatsbyImage
        alt={title}
        image={image}
        css={css`
          position: initial;
        `}
      />
      <h2 css={postHeaderStyle}>{title}</h2>
    </div>
  );
};

export default CoverImage;
