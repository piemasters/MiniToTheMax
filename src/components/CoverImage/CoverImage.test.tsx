import { getByTestId } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';

import { CoverImage } from './CoverImage';
import { appTheme } from '../../styles/theme';
import { renderWithTransitionProvider } from '../../util';

describe('CoverImage', () => {
  test('renders correctly', () => {
    const { container } = renderWithTransitionProvider(
      <ThemeProvider theme={appTheme}>
        <CoverImage
          image={{
            layout: 'constrained',
            width: 1800,
            height: 900,
            backgroundColor: '#282828',
            images: {
              fallback: {
                src: 'cover.jpeg',
              },
            },
          }}
          title={'Demo Post Title'}
        />
      </ThemeProvider>
    );

    expect(getByTestId(container, 'cover-image')).toHaveTextContent(
      'Demo Post Title'
    );
    expect(container).toMatchSnapshot();
  });
});
