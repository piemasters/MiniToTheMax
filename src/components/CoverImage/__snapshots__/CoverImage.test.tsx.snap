// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoverImage renders correctly 1`] = `
<div>
  <div
    css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    data-testid="cover-image"
  >
    <div
      class="gatsby-image-wrapper gatsby-image-wrapper-constrained"
      data-gatsby-image-wrapper=""
      style="position: relative; overflow: hidden; display: inline-block; vertical-align: top;"
    >
      <div
        style="max-width: 1800px; display: block;"
      >
        <img
          alt=""
          aria-hidden="true"
          role="presentation"
          src="data:image/svg+xml;charset=utf-8,%3Csvg%20height='900'%20width='1800'%20xmlns='http://www.w3.org/2000/svg'%20version='1.1'%3E%3C/svg%3E"
          style="max-width: 100%; display: block; position: static;"
        />
      </div>
      <div
        aria-hidden="true"
        data-placeholder-image=""
        style="height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
      />
      <img
        alt="Demo Post Title"
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        data-gatsby-image-ssr=""
        data-main-image=""
        data-src="cover.jpeg"
        decoding="async"
        loading="lazy"
        style="height: 100%; left: 0px; position: absolute; top: 0px; transform: translateZ(0); transition: opacity 250ms linear; width: 100%; will-change: opacity; opacity: 0;"
      />
      <noscript />
      <script
        type="module"
      >
        const t="undefined"!=typeof HTMLImageElement&&"loading"in HTMLImageElement.prototype;if(t){const t=document.querySelectorAll("img[data-main-image]");for(let e of t){e.dataset.src&&(e.setAttribute("src",e.dataset.src),e.removeAttribute("data-src")),e.dataset.srcset&&(e.setAttribute("srcset",e.dataset.srcset),e.removeAttribute("data-srcset"));const t=e.parentNode.querySelectorAll("source[data-srcset]");for(let e of t)e.setAttribute("srcset",e.dataset.srcset),e.removeAttribute("data-srcset");e.complete&&(e.style.opacity=1,e.parentNode.parentNode.querySelector("[data-placeholder-image]").style.opacity=0)}}
      </script>
    </div>
    <h2
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    >
      Demo Post Title
    </h2>
  </div>
</div>
`;
